website/static/js/addProjectPlugin.js

```diff
@@ -0,0 +1,178 @@
+/**
+ * UI and function to add project
+ */
+'use strict';
+
+require('css/add-project-plugin.css');
+var $ = require('jquery');
+var m = require('mithril');
+var $osf = require('js/osfHelpers');
+
+// XHR configuration to get apiserver connection to work
+var xhrconfig = function (xhr) {
+    xhr.withCredentials = true;
+};
+
+
+var AddProject = {
+    controller : function (options) {
+        var self = this;
+        self.defaults = {
+            buttonTemplate : m('.btn.btn-primary[data-toggle="modal"][data-target="#addProjectModal"]', 'Add new Project'),
+            parent : null
+        };
+        self.viewState = m.prop('form'); // 'processing', 'success', 'error';
+        self.options = $.extend(self.defaults, options);
+        self.init = function _init () {
+
+        };
+        self.newProjectName = m.prop('');
+        self.newProjectDesc = m.prop('');
+        self.newProjectCategory = 'project';
+        self.goToProjectLink = m.prop('');
+        self.errorMessageType = m.prop('unknown');
+        self.errorMessage = {
+            'unknown' : 'There was an unknown error. Please try again later.'
+        };
+        self.chooseCategory = function(event){
+            self.newProjectCategory = $(this).val();
+        };
+        self.add = function _add () {
+            self.viewState('processing');
+            var url = $osf.apiV2Url('nodes/', { query : {}});
+            var data = {
+                    'data' : {
+                        'type': 'nodes',
+                        'attributes': {
+                            'title': self.newProjectName(),
+                            'category': self.newProjectCategory,
+                            'description' : self.newProjectDesc()
+                        }
+                    }
+                };
+            var success = function _success (result) {
+                self.viewState('success');
+                console.log('success', result);
+                self.goToProjectLink(result.data.links.html);
+            };
+            var error = function _error (result) {
+                self.viewState('error');
+                console.log('error', result);
+            };
+            m.request({method : 'POST', url : url, data : data, config : xhrconfig})
+                .then(success, error);
+            self.newProjectName('');
+            m.redraw(true);
+
```

Comment:
so something like this:
```
self.add = function() {
    ....  
    return m.request(...)
}

self.add().then(success, error)
```
I don't see the benefit of doing this in this case though, and it seems to be decoupling success and error responses which are still part of adding. Can you write out an example of the refactoring?